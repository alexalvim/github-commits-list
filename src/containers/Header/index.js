import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Autocomplete from 'react-autocomplete';

import UserListItem from '../../components/UserListItem'
import logo from '../../imgs/light-logo.png';
import {
  searchUserRequest,
  getUserRepositoriesRequest,
  getUserRequest
} from '../../actions/user';
import {
  ContentWrapper,
  Container,
  Logo,
  autocompleteInputStyle,
  autocompleteWrapperStyle
} from './styles';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userField: '',
    }
  }

  handleChangeUserInput = (evt) => {
    const { searchUserRequest } = this.props;
    const user = evt.currentTarget.value;
    this.setState({
      userField: user
    });
    if(user.length > 0) {
      searchUserRequest(user);
    }
  }

  handleOnClickUserListItem = (login, avatarUrl, evt) => {
    const { getUserRepositoriesRequest, getUserRequest } = this.props;
    this.setState({
      userField: ''
    });
    getUserRepositoriesRequest(login, avatarUrl, 1);
    getUserRequest(login);
  }

  render() {
    const { searchedUsers } = this.props;
    const { userField } = this.state;
    const { handleChangeUserInput, handleOnClickUserListItem } = this;
    return (
      <ContentWrapper>
        <Container>
          <Logo src={logo} />
          <Autocomplete
            getItemValue={(user) => user.login}
            items={searchedUsers}
            renderItem={(user) =>
              <div key={user.id}>
                <UserListItem
                 title={user.login}
                 imgUrl={user.avatar_url}/>
              </div>
            }
            value={userField}
            onChange={handleChangeUserInput}
            inputProps={{...autocompleteInputStyle, placeholder: 'Digite o usuÃ¡rio a ser buscado'}}
            wrapperStyle={autocompleteWrapperStyle}
            onSelect={(_, user) => handleOnClickUserListItem(user.login, user.avatar_url)}
          />
        </Container>
      </ContentWrapper>
    );
  }
}

const mapStateToProps = store => ({
  searchedUsers: store.userState.searchedUsers,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ searchUserRequest, getUserRepositoriesRequest, getUserRequest }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Header);
