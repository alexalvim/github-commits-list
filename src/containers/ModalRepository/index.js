import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import InfiniteScroll from 'react-infinite-scroll-component';
import { IoMdSearch } from 'react-icons/io'

import {
  clearRepository,
  searchCommitRequest,
  getRepositoryCommitsRequest
} from '../../actions/repository';
import SimpleItem from '../../components/SimpleItem';
import ModalBox from '../../components/ModalBox';
import TextField from '../../components/TextField';
import TextButton from '../../components/TextButton';
import IconButton from '../../components/IconButton'
import {
  CommitsList,
  CommitsListWrapper,
  CommitsForm,
  commitTextFieldStyle,
  clearFilterButtonStyle
} from './styles';

class ModalRepository extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      commitSearchField: '',
      showClearFilter: false
    }
  }

  handleSubmitCommitSearch = (evt) => {
    evt.preventDefault();
    const { searchCommitRequest, user, repository } = this.props;
    const { commitSearchField } = this.state;
    if(commitSearchField.trim().length > 0) {
      searchCommitRequest(user.login, repository.name, commitSearchField);
      this.setState({
        showClearFilter: true
      });
    }
  }

  handleCloseModal = () => {
    const { closeModal, clearRepository } = this.props;
    this.setState({
      commitSearchField: '',
    })
    clearRepository();
    closeModal()
  }

  handleChangeCommitSearchField = (evt) => {
    const value = evt.currentTarget.value;
    this.setState({
      commitSearchField: value
    });
  }

  handleClearCommitFilters = () => {
    const { user, repository, getRepositoryCommitsRequest } = this.props;
    getRepositoryCommitsRequest(user.login, repository.name, 1)
    this.setState({
      showClearFilter: false,
      commitSearchField: ''
    });
  }

  handleSeeMoreCommits = () => {
    const { user, repository, getRepositoryCommitsRequest } = this.props;
    getRepositoryCommitsRequest(user.login, repository.name, repository.commitsPage + 1);
  }

  render() {
    const { repository, isOpen } = this.props
    const { commitSearchField, showClearFilter } = this.state;
    const {
      handleCloseModal,
      handleSubmitCommitSearch,
      handleChangeCommitSearchField,
      handleClearCommitFilters,
      handleSeeMoreCommits
    } = this;
    return (
      <ModalBox
        title={repository.name}
        isOpen={isOpen}
        closeModal={handleCloseModal}>
        <Fragment>
          <CommitsForm
            onSubmit={handleSubmitCommitSearch}>
            <TextField
              value={commitSearchField}
              placeholder='Digite aqui um termo para filtrar'
              onChange={handleChangeCommitSearchField}
              wrapperStyle={commitTextFieldStyle}/>
            <IconButton
              onClick={handleSubmitCommitSearch}
              Icon={IoMdSearch}/>
          </CommitsForm>
          {showClearFilter &&
            <TextButton
              wrapperStyle={clearFilterButtonStyle}
              onClick={handleClearCommitFilters}
              label='Limpar Filtros'/>}
          {repository.isLoadingCommits ?
            <span>Carregando Commits...</span>:
            repository.commits && repository.commits.length > 0 ?
              <CommitsListWrapper id='commits-list-wrapper'>
                <InfiniteScroll
                  scrollableTarget='commits-list-wrapper'
                  dataLength={repository.commits && repository.commits.length}
                  next={handleSeeMoreCommits}
                  hasMore={repository.hasNextPage}
                  loader={<span>Carregando Commits...</span>}>
                  <CommitsList>
                    {repository.commits.map((commit) => 
                      <li key={commit.sha}>
                      <SimpleItem
                        title={commit.commit.message}
                        description={commit.sha}/>
                      </li>)}
                  </CommitsList>
                </InfiniteScroll>
              </CommitsListWrapper>
            : <span>Commits n√£o encontrados</span>
          }
        </Fragment>
      </ModalBox>
    )
  }
}

const mapStateToProps = store => ({
  repository: store.repositoryState,
  user: store.userState,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({
    clearRepository,
    searchCommitRequest,
    getRepositoryCommitsRequest
  }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ModalRepository);