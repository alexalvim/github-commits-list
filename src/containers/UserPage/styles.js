import styled from 'styled-components';

import Colors from '../../colors';
import Spaces from '../../spaces';

const Container = styled.div`
  margin: 0 auto;
  max-width: ${Spaces.containerMaxWidth};
  padding: ${Spaces.double};
`;

const ContentHolder = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: column;

  @media (min-width: ${Spaces.tabletBreakpoint}) {
    flex-direction: row;
  }
`;

const RepositoriesList = styled.ul`
  > li {
    border-bottom: solid 1px ${Colors.lightGray};

    &:last-child {
      border: none;
    }
  }
`;

const CommitsList = styled.ul`
  overflow: auto;
  padding-right: ${Spaces.base};

  > li {
    word-break: break-all;
  }
`;

const contentBoxWrapperStyle = `
  margin-top: ${Spaces.double};

  @media (min-width: ${Spaces.tabletBreakpoint}) {
    margin-left: ${Spaces.double};
    margin-top: 0;
  }
`;

export {
  Container,
  ContentHolder,
  RepositoriesList,
  CommitsList,
  contentBoxWrapperStyle
}