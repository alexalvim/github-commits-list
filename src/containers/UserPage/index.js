import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import Header from '../Header';
import UserBox from '../../components/UserBox';
import ContentBox from '../../components/ContentBox';
import SimpleItem from '../../components/SimpleItem';
import { searchUserRequest, userRepositoriesRequest } from '../../actions/user';
import {
  Container,
  ContentHolder,
  RepositoriesList,
  contentBoxWrapperStyle
} from './styles';

class UserPage extends React.Component {

  handleClickSeeMore = () => {
    const { user, userRepositoriesRequest } = this.props;
    userRepositoriesRequest(user.login, user.avatarUrl, user.repositoriesPage + 1);
  }

  render() {
    const { user } = this.props;
    const { handleClickSeeMore } = this;
    console.log(user)
    return (
      <Fragment>
        <Header/>
        <Container>
          {user.login && user.login !== '' ?
          <ContentHolder>
            <UserBox
              avatarUrl={user.avatarUrl}
              login={user.login}/>
            <ContentBox
             title='Repositórios'
             wrapperStyles={contentBoxWrapperStyle}>
               {user.repositories &&
                user.repositories.length > 0 ?
                <Fragment>
                  <RepositoriesList>
                    {
                     user.repositories.map((repository) =>
                      <li key={repository.id}>
                        <SimpleItem
                          title={repository.name}
                          description={repository.description}/>
                      </li>
                    )}
                  </RepositoriesList>
                  {user.repositories.length < user.repositoriesCount &&
                    <span onClick={handleClickSeeMore}>ver mais</span>}
                </Fragment>
                : <p>Usuário não possui repositórios</p>}
            </ContentBox>
          </ContentHolder>
          : <p>Aplicação para verificar os repositórios e commits de usuários do Github. <br/>Use o campo acima para buscar um usuário</p>}
        </Container>
      </Fragment>
    );
  }
}

const mapStateToProps = store => ({
  user: store.userState,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ searchUserRequest, userRepositoriesRequest }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(UserPage);
