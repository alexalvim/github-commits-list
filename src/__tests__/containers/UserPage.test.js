import React from 'react';
import { shallow } from 'enzyme';

import { UserPage } from '../../containers/UserPage';

const user = {
  login: 'loginTest',
  avatarUrl: 'avatarUrlTest',
  repositoriesPage: 1
}

const repository = {
  name: 'repositoryNameTest'
}

describe('Testing UserPage', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<UserPage user={user} repository={repository} />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Testing UserPage handle functions', () => {
  it('should call getUserRepositoriesRequest when call handleSeeMoreRepositories', () => {
    const mockFunction = jest.fn();
    const wrapper = shallow(
      <UserPage
        getUserRepositoriesRequest={mockFunction}
        user={user}
        repository={repository} />);

    wrapper.instance().handleSeeMoreRepositories();
    expect(mockFunction).toHaveBeenCalledWith(user.login, user.avatarUrl, user.repositoriesPage + 1);
  });

  it('should change the state value and call getRepositoryCommitsRequest when call handleOnClickRepository', () => {
    const mockFunction = jest.fn()
    const repositoryName = 'name'
    const wrapper = shallow(
      <UserPage
        getRepositoryCommitsRequest={mockFunction}
        user={user}
        repository={repository} />);

    wrapper.instance().state.openRepositoryModal = false;
    wrapper.instance().handleOnClickRepository(repositoryName);
    expect(wrapper.instance().state.openRepositoryModal).toEqual(true);
    expect(mockFunction).toHaveBeenCalledWith(user.login, repositoryName, 1);
  });

  it('should change state value when call handleCloseModalRepository', () => {
    const wrapper = shallow(
      <UserPage
        user={user}
        repository={repository} />);

    wrapper.instance().state.openRepositoryModal = true
    wrapper.instance().handleCloseModalRepository();
    expect(wrapper.instance().state.openRepositoryModal).toEqual(false);
  });

  it('should call clearErrorMessage when call handleCloseModalRepository', () => {
    const mockFunction = jest.fn();
    const wrapper = shallow(
      <UserPage
        clearErrorMessage={mockFunction}
        user={user}
        repository={repository} />);

    wrapper.instance().handleCloseModalError();
    expect(mockFunction).toHaveBeenCalled()
  });
});