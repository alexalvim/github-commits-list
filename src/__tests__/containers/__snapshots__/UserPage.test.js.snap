// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserPage should render correctly 1`] = `ShallowWrapper {}`;
